SerialPrograms/Source/PokemonSV/Inference/ItemPrinter/PokemonSV_ItemPrinterMaterialDetector.cpp

```diff
@@ -0,0 +1,289 @@
+int8_t ItemPrinterMaterialDetector::find_happiny_dust_row_index(
+    AsyncDispatcher& dispatcher,
+    ConsoleHandle& console, BotBaseContext& context
+) const{
+    int8_t value_68_row_index;
+    for (size_t c = 0; c < 10; c++){
+        context.wait_for_all_requests();
+        value_68_row_index = find_material_value_row_index(dispatcher, console, context, 68);
+        if (value_68_row_index != -1 
+            && detect_material_name(console, context, value_68_row_index) == "happiny-dust"
+        ){  
+            // found screen and row number with Happiny dust.
+            // std::cout << "Happiny dust found. Row number: " << std::to_string(value_68_row_index) << std::endl;
+            return value_68_row_index;
+        }
+        // keep searching for Happiny dust
+        pbf_press_dpad(context, DPAD_RIGHT, 20, 30);
+    }
+
+    throw OperationFailedException(
+        ErrorReport::SEND_ERROR_REPORT,
+        console,
+        "Failed to find Happiny dust after 10 tries."
+    );
+
+}
+
+// detects the material name at the given row_index
+// MaterialNameReader has a very limited dictionary,
+// so it can only reliably read the material names with 68% value
+// (i.e. Ditto Goo, Happiny Dust, Magby Hair, Beldum Claw)
+std::string ItemPrinterMaterialDetector::detect_material_name(
+    ConsoleHandle& console, 
+    BotBaseContext& context,
+    int8_t row_index
+) const{
+    VideoSnapshot snapshot = console.video().snapshot();
+    ImageFloatBox material_name_box = m_box_mat_name[row_index];
+    ImageViewRGB32 material_name_image = extract_box_reference(snapshot, material_name_box);
+    const auto ocr_result = MaterialNameReader::instance().read_substring(
+        console, m_language, 
+        material_name_image, OCR::BLACK_OR_WHITE_TEXT_FILTERS()
+    );
+
+    std::multimap<double, OCR::StringMatchData> results;
+    if (!ocr_result.results.empty()){
+        for (const auto& result : ocr_result.results){
+            results.emplace(result.first, result.second);
+        }
+    }
+
+    if (results.empty()){
+        return "";
+    }
+
+    if (results.size() > 1){
+        throw OperationFailedException(
+            ErrorReport::SEND_ERROR_REPORT, 
+            console,
+            "ItemPrinterMaterialDetector::detect_material_name(): Unable to read selected item. Ambiguous or multiple results."
+        );
+    }
+
+    return results.begin()->second.token;
+}
+
+// return row number that matches given material_value. else return -1
+int8_t ItemPrinterMaterialDetector::find_material_value_row_index(
+    AsyncDispatcher& dispatcher,
+    ConsoleHandle& console, 
+    BotBaseContext& context,
+    int16_t material_value
+) const{
+    context.wait_for_all_requests();
+    VideoSnapshot snapshot = console.video().snapshot();
+    for (int8_t i = 0; i < 10; i++){
+        int16_t value = read_number(console, dispatcher, snapshot, m_box_mat_value[i]);
+        if (value == material_value){
+            return i;
```

Comment:
What happens if there are multiple 68% materials on the same screen?

Reply:
My bad. I keep forgetting that there can be missing materials from the list. I'll make the function return a vector instead.